testers.lycheeLinkCheck: init

roberth opened the pull request:
## Description of changes

- Add a helper to perform an offline, internal link check.

- Add passthru `tests` to the `lychee` package.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
testers.lycheeLinkCheck: init

roberth committed:
testers.lycheeLinkCheck: Add online attribute for nix run

roberth committed:
doc/testers: Make lycheeLinkCheck section more accessible

- Explain the use case for audience who's new to the concept
- Be more explicit about what the return value "does"

roberth commented:
@ofborg eval

totoroot reviewed (approved):
Thank you, @roberth! Especially for the terrific documentation.

stephen-huan reviewed (approved):
+1. I'd love to see this merged.

roberth committed:
testers.lycheeLinkCheck: Test `online` runner

roberth commented:
@ofborg build lychee.tests

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3969

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -36,6 +36,61 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+
+Check a packaged static site's links with the [`lychee` package](https://search.nixos.org/packages?show=lychee&type=packages&query=lychee).
+
+You may use Nix to reproducibly build documentation sites and other static sites. Some packages will install documentation in their `out` or `doc` outputs, or maybe you have dedicated package where you've made your static site reproducible by running a generator (like [Hugo](https://gohugo.io/), or [mdBook](https://rust-lang.github.io/mdBook/)) in a derivation.
```
```suggestion
You may use Nix to reproducibly build documentation sites and other static sites.
Some packages will install documentation in their `out` or `doc` outputs, or maybe you have dedicated package where you've made your static site reproducible by running a generator, such as [Hugo](https://gohugo.io/) or [mdBook](https://rust-lang.github.io/mdBook/), in a derivation.
```

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -36,6 +36,61 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+
+You may use Nix to reproducibly build documentation sites and other static sites. Some packages will install documentation in their `out` or `doc` outputs, or maybe you have dedicated package where you've made your static site reproducible by running a generator (like [Hugo](https://gohugo.io/), or [mdBook](https://rust-lang.github.io/mdBook/)) in a derivation.
+
+If you have a static site that is buildable in Nix, you can use `lycheeLinkCheck` to check that the hyperlinks in your site are correct, and do so as part of your Nix workflow and CI.
```
```suggestion
If you have a static site that can be built with Nix, you can use `lycheeLinkCheck` to check that the hyperlinks in your site are correct, and do so as part of your Nix workflow and CI.
```

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -36,6 +36,61 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+
+This tester produces a package that does not produce useful outputs, but only succeeds if the hyperlinks in your site are correct. The build log will list the broken links.
+
+You may use it in two modes:
```
```suggestion
It has two modes:
```

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -36,6 +36,61 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+
+- Use the returned package directly, and the build process will check that internal hyperlinks are correct. This runs in the sandbox, so it will not check external hyperlinks, but it is quick and reliable.
+
+- Invoke the `.online` with [`nix run`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-run) ([experimental](https://nixos.org/manual/nix/stable/contributing/experimental-features#xp-feature-nix-command)). This runs outside the sandbox, and check that bith internal and external hyperlinks are correct. Example:
```
```suggestion
- Invoke the `.online` attribute with [`nix run`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-run) ([experimental](https://nixos.org/manual/nix/stable/contributing/experimental-features#xp-feature-nix-command)). This runs outside the sandbox, and check that both internal and external hyperlinks are correct.
  Example:
```

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
````diff
@@ -36,6 +36,61 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+  nix run nixpkgs#lychee.tests.ok.online
+  ```
+
+### Parameters {#tester-lycheeLinkCheck-params}
+
+#### `site` (required) {#tester-lycheeLinkCheck-param-site}
+
+The path to the files to check.
+
+#### `extraConfig` (optional) {#tester-lycheeLinkCheck-param-extraConfig}
+
+Extra configuration to pass to `lychee` in its [config file](https://github.com/lycheeverse/lychee/blob/master/lychee.example.toml). It is automatically [translated](https://nixos.org/manual/nixos/stable/index.html#sec-settings-nix-representable) to TOML.
+
+Example: `{ "include_verbatim" = true; }`
+
+#### `lychee` (optional) {#tester-lycheeLinkCheck-param-lychee}
+
+The `lychee` package to use.
+
+#### `remapUrl` (optional) {#tester-lycheeLinkCheck-param-remapUrl}
+
+A URL to which the files may be deployed.
+Instead of ignoring this URL because the check is offline, `lychee` will check that the links to this URL are correct, by pretending `site` is deployed there.
+
````
Per [conventions](https://github.com/NixOS/nixpkgs/blob/master/doc/README.md#documentation-conventions) we now use definition lists for parameter listings:

```suggestion
### Inputs {#tester-lycheeLinkCheck-inputs}

`site` (path or derivation) {#tester-lycheeLinkCheck-param-site}

: The path to the files to check.

`extraConfig` (attribute set) {#tester-lycheeLinkCheck-param-extraConfig}

: Extra configuration to pass to `lychee` in its [configuration file](https://github.com/lycheeverse/lychee/blob/master/lychee.example.toml).
  It is automatically [translated](https://nixos.org/manual/nixos/stable/index.html#sec-settings-nix-representable) to TOML.

  Example: `{ "include_verbatim" = true; }`

`lychee` (derivation, optional) {#tester-lycheeLinkCheck-param-lychee}

: The `lychee` package to use.
```

fricklerhandwerk commented on pkgs/build-support/testers/lychee.nix:
```diff
@@ -0,0 +1,49 @@
+  # or doc/builders/testers.chapter.md
+  lycheeLinkCheck = {
+    site,
+    remapUrl ? null,
```
```suggestion
```

roberth replied:
I've replaced it by the `remap` argument, which is more general, correctly documented (regex!) and about as easy to use for the original use case of `remapUrl` but without catching users in an overly specific solution.

fricklerhandwerk commented on pkgs/build-support/testers/lychee.nix:
```diff
@@ -0,0 +1,49 @@
+
+      # These can be overriden with overrideAttrs if needed.
+      passthru = {
+        inherit lychee remapUrl;
```
```suggestion
        inherit lychee;
```

fricklerhandwerk commented on pkgs/build-support/testers/lychee.nix:
```diff
@@ -0,0 +1,49 @@
+        inherit lychee remapUrl;
+        config = {
+          include_fragments = true;
+        } // lib.optionalAttrs (finalAttrs.passthru.remapUrl != null) {
+          remap = [ "${remapUrl} file://${finalAttrs.site}" ];
```
```suggestion
```

fricklerhandwerk reviewed (approved):
Reviewed together with @roberth at #ZurichZHF. I suggest to remove the `remapUrl` feature, because it's unnecessarily constraining remapping to the same site. Already for linking between the Nix/Nixpkgs/NixOS manuals I'd like to configure where these remapped URL contents can be found, and those should be different trees. I agree though that expressing this as `extraConfig` is tedious, and will merge a PR that models remapping appropriately at the Nix language layer.

PS: Met a @thomas-zahner (co-maintainer of lychee) today and exchanged ideas! Thanks for the great day.

roberth committed:
testers.lycheeLinkCheck: Edit documentation

Co-authored-by: Valentin Gagarin <valentin.gagarin@tweag.io>

roberth committed:
testers.lycheeLinkCheck: Generalize remapUrl convenience feature

roberth committed:
testers.lycheeLinkCheck: Update docs

A deep dive reveals that the needles in our haystack are actually
regular expressions.

roberth committed:
testers.lycheeLinkCheck: Fix plain derivation remap, showcase emptyDirectory

roberth committed:
testers.lycheeLinkCheck: Validate regex...

... because of the odd choice of config syntax.

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -36,6 +36,82 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+
+Check a packaged static site's links with the [`lychee` package](https://search.nixos.org/packages?show=lychee&type=packages&query=lychee).
+
+You may use Nix to reproducibly build documentation sites and other static sites.
```
```suggestion
You may use Nix to reproducibly build static websites, such as for software documentation.
```

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -36,6 +36,82 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+`remap` (attribe set, optional) {#tester-lycheeLinkCheck-param-remap}
+
+: An attribute set where the attribute names are regular expressions.
+  The values should be strings, derivations or path values.
```
```suggestion
  The values should be strings, derivations, or path values.
```

We seem to have converged on the Oxford comma. (Maybe I should write that into the guidelines...)

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -36,6 +36,82 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+
+  In the returned check's default configuration, external URLs are only checked when you run the `.online` attribute.
+
+  By adding remappings, you can check offline that URLs to external resources are correct, by providing a file system based stand-in.
```
```suggestion
  By adding remappings, you can check offline that URLs to external resources are correct, by providing a stand-in from the file system.
```

fricklerhandwerk commented on doc/build-helpers/testers.chapter.md:
````diff
@@ -36,6 +36,82 @@ passthru.tests.pkg-config = testers.hasPkgConfigModules {
+  }
+  ```
+
+  Store path in the attribute values are automatically prefixed with `file://`, because lychee requires this for paths in the file system.
````
```suggestion
  Store paths in the attribute values are automatically prefixed with `file://`, because lychee requires this for paths in the file system.
```

fricklerhandwerk commented on pkgs/build-support/testers/lychee.nix:
```diff
@@ -0,0 +1,66 @@
+    else s;
+
+  withCheckedName = name:
+    throwIf
+      (hasInfix " " name)
+      "lycheeLinkCheck: remap patterns must not contain spaces. A space marks the end of the regex in lychee.toml. Please change attribute name remap.${escapeNixString name}";
```
```suggestion
    throwIf (hasInfix " " name) ''
      lycheeLinkCheck: remap patterns must not contain spaces.
      A space marks the end of the regex in lychee.toml.

      Please change attribute name 'remap.${escapeNixString name}'
    '';
```

fricklerhandwerk commented on pkgs/tools/networking/lychee/tests/fail.nix:
```diff
@@ -0,0 +1,21 @@
+
+in
+  runCommand "link-check-fail" { inherit failure; } ''
+    grep -F butwhere $failure/testBuildFailure.log >/dev/null
```
What's `butwhere`??

roberth replied:
Changed it to `foos-missing-anchor`
`missing-anchor` would have been nicer but could accidentally match an error code or something.

fricklerhandwerk reviewed (approved):
Awesome tests, too!

roberth committed:
testers.lycheeLinkCheck: Edit documentation and messages

Co-authored-by: Valentin Gagarin <valentin.gagarin@tweag.io>

roberth committed:
testers.lycheeLinkCheck: Clarify test data
